lsp.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2022 Mineiros GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tmlsp
+
+import (
+	"context"
+	"encoding/json"
+	"log"
+
+	"go.lsp.dev/jsonrpc2"
+	lsp "go.lsp.dev/protocol"
+	"go.lsp.dev/uri"
+)
+
+type Server struct {
+	conn jsonrpc2.Conn
+
+	workspace string
+}
+
+type initParams struct {
+	ProcessID int    `json:"processId,omitempty"`
+	RootURI   string `json:"rootUri,omitempty"`
+}
+
+func NewServer(conn jsonrpc2.Conn) *Server {
+	return &Server{
+		conn: conn,
+	}
+}
+
+// Handler handles the client requests.
+func (s *Server) Handler(ctx context.Context, reply jsonrpc2.Replier, r jsonrpc2.Request) error {
+	log.Printf("got request %s with params: %s", r.Method(), r.Params())
+	switch r.Method() {
+	default:
+		log.Printf("%s is not implemented", r.Method())
+	case lsp.MethodInitialize:
+		var params initParams
+		if err := json.Unmarshal(r.Params(), &params); err != nil {
+			log.Fatal(err)
+		}
+
+		s.workspace = string(uri.New(params.RootURI).Filename())
+		err := reply(ctx, lsp.InitializeResult{
+			Capabilities: lsp.ServerCapabilities{
+				CompletionProvider: &lsp.CompletionOptions{},
+
+				// if we support `goto` definition.
+				DefinitionProvider: false,
+
+				// If we support `hover` info.
+				HoverProvider: false,
+
+				TextDocumentSync: lsp.TextDocumentSyncOptions{
+					// Send all file content on every change (can be optimized later).
+					Change: lsp.TextDocumentSyncKindFull,
+
+					// if we want to be notified about open/close of Terramate files.
+					OpenClose: true,
+					Save: &lsp.SaveOptions{
+						// If we want the file content on save,
+						IncludeText: false,
+					},
+				},
+			},
+		}, nil)
+
+		if err != nil {
```

Comment:
crashing the server if we can't send a response is the way to go here ? I imagine there can be more than one client connected on the server on some setups, but disregard if this is premature/stupid (never worked with jsonrpc servers).

Reply:
In `stdio` mode there's only one client, the vscode workspace window managing multiple files, or the file session in vim/emacs/etc, then it's always one client connected. I guess this way the jsonrpc could only fail if the server cannot read/write to stdio/stdout... ie the parent process cannot read/write to the child, so if it fails, there's not much we can do.

This method makes the design very simple, but probably not the best option in the long run but ... it does seem like lots of extensions still use it by default even `gopls` in the official vscode-go [see here](https://github.com/golang/vscode-go/blob/master/package.json#L1541) and [here](https://github.com/golang/tools/blob/master/gopls/doc/daemon.md). The gopls docs about running it in a daemon talk about some drawbacks of the stdio mode and also explain the quite intricate forwarders and daemon involved. I think we should study what they did there and maybe copy the approach.

Reply:
hmm if we can push for always having a server per client even when working on other modes this would be fine, was just wondering how the other modes would work etc, but for now we only support stdio so it does make sense that aborting is fine in this case.